### PR TITLE
Q2ra/fiyaz bin hasan/add q/2

### PR DESCRIPTION
**Pull request by Rick-Anderson**

### CODE REVIEW

**Comment by Rick-Anderson** [comment]
@fiyazbinhasan I moved your commit over here. Can we get rid of  `EmailService` ? 

I think the sample is too complex. All we want to show is `/Send` and a background task, each that send a simulated email.

**Comment by fiyazbinhasan** [comment]
I can give it a try but I'm afraid it will fall back into the initial sample that I proposed

**Comment by fiyazbinhasan** [comment]
Hi @Rick-Anderson, I might not be able to pick this up before the end of the week. Just letting you know

**Comment by fiyazbinhasan** [comment]
Hi @Rick-Anderson, sorry for getting back late to this issue. However, I was thinking that this sample is way too complicated just to show the caveats of HttpContext. I came up with a new sample which is a bit easier to understand. It will read the user-agent string from the request before making a request to the github api. If the request is without a httpcontext, it will append the unknown string to the header. Let me know what you think. If you give me a go, I will update the code snippet and commit.

```
using System.Text.Json;
using HttpContextInBackgroundThread;
using Microsoft.Net.Http.Headers;

var builder = WebApplication.CreateBuilder(args);

builder.Services.AddHttpContextAccessor();
builder.Services.AddSingleton<BranchesLoggerService>();

builder.Services.AddHttpClient("GitHub", httpClient =>
{
    httpClient.BaseAddress = new Uri("https://api.github.com/");

    // using Microsoft.Net.Http.Headers;
    // The GitHub API requires two headers. The Use-Agent header will be added dynamically through UserAgentHeaderHandler
    httpClient.DefaultRequestHeaders.Add(
        HeaderNames.Accept, "application/vnd.github.v3+json");
}).AddHttpMessageHandler<UserAgentHeaderHandler>();

builder.Services.AddTransient<UserAgentHeaderHandler>();

var app = builder.Build();

app.MapGet("/", () => "Hello World!");

app.MapGet("/branches", async (IHttpClientFactory httpClientFactory) =>
{
    var httpClient = httpClientFactory.CreateClient("GitHub");
    var httpResponseMessage = await httpClient.GetAsync(
        "repos/dotnet/AspNetCore.Docs/branches");

    if (httpResponseMessage.IsSuccessStatusCode)
    {
        using var contentStream =
            await httpResponseMessage.Content.ReadAsStreamAsync();

        return Results.Ok(await JsonSerializer.DeserializeAsync
            <IEnumerable<GitHubBranch>>(contentStream));
    } else
    {
        return Results.BadRequest();
    }
});

var branchesLogger = app.Services.GetService<BranchesLoggerService>();
await branchesLogger!.LogBranchesAsync();

app.Run();


public class BranchesLoggerService
{

    public readonly IHttpClientFactory _httpClientFactor;
    public readonly ILogger<BranchesLoggerService> _logger;

    public BranchesLoggerService(IHttpClientFactory httpClientFactory, ILogger<BranchesLoggerService> logger)
    {
        _httpClientFactor = httpClientFactory;
        _logger = logger;
    }

    public async Task LogBranchesAsync()
    {
        var httpClient = _httpClientFactor.CreateClient("GitHub");
        var httpResponseMessage = await httpClient.GetAsync(
            "repos/dotnet/AspNetCore.Docs/branches");

        if (httpResponseMessage.IsSuccessStatusCode)
        {
            var contentString = await httpResponseMessage.Content.ReadAsStringAsync();

            _logger.LogInformation(contentString);
        }
    }
}

public class UserAgentHeaderHandler : DelegatingHandler
{
    private readonly IHttpContextAccessor _httpContextAccessor;

    public UserAgentHeaderHandler(IHttpContextAccessor httpContextAccessor)
    {
        _httpContextAccessor = httpContextAccessor;
    }

    protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
    {
        var contextRequest = _httpContextAccessor.HttpContext?.Request;
        string? userAgentString = contextRequest?.Headers["user-agent"].ToString();

        if (string.IsNullOrEmpty(userAgentString))
        {
            userAgentString = "Unknown";
        }

        request.Headers.Add(HeaderNames.UserAgent, userAgentString);

        return await base.SendAsync(request, cancellationToken);
    }
}
```

**Comment by Rick-Anderson** [comment]
@halter73 please review preceding sample ***IN THE COMMENTS ABOVE*** to see if that sample will work for issue https://github.com/dotnet/AspNetCore.Docs/issues/7721 and for PR #25701

**Comment by halter73** [comment]
I think what @fiyazbinhasan proposed is pretty good. I think we still want to demonstrate using the service (in this case, the `BranchesLoggerService`) from a request (where the `_httpContextAccessor.HttpContext` is set) and from a `BackgroundService`  (where the `_httpContextAccessor.HttpContext` is nost set). I don't want to show of `app.Services.GetService()` more than necessary. In the `BackgroundService` case, the `BackgroundService` should manually set `request.Headers.Add(HeaderNames.UserAgent, ...);` to something. It should also show how to gracefully handle errors and shutdown like https://github.com/dotnet/AspNetCore.Docs/pull/25701#discussion_r890716183 does. 

Maybe the `BackgroundService` could be called the `BranchesLoggerService`. I'm not sure why you want to log github branches in the background or during the request for that matter, but I suppose all these samples are somewhat contrived. Emails and newsletters did seem a little more real world though.

**Comment by fiyazbinhasan** [comment]
@Rick-Anderson not sure how to commit changes to this pull request. Should I create another pull request on my forked repo against this pull request branch?

File: aspnetcore/fundamentals/http-context/samples/6.x/HttpContextInBackgroundThread/Program.cs [hunk 1]
```diff
@@ -1,13 +1,45 @@
...
+    {
+        using var contentStream =
+            await httpResponseMessage.Content.ReadAsStreamAsync();
+
+        return Results.Ok(await JsonSerializer.DeserializeAsync
+            <IEnumerable<GitHubBranch>>(contentStream));
+    } else
+    {
+        return Results.BadRequest();
+    }
```
Comment (fiyazbinhasan) [inline aspnetcore/fundamentals/http-context/samples/6.x/HttpContextInBackgroundThread/Program.cs:42]:
```suggestion
    var httpClient = httpClientFactory.CreateClient("GitHub");
    var httpResponseMessage = await httpClient.GetAsync(
        "repos/dotnet/AspNetCore.Docs/branches");

    if (!httpResponseMessage.IsSuccessStatusCode) 
        return Results.BadRequest();

    await using var contentStream =
        await httpResponseMessage.Content.ReadAsStreamAsync();

    return Results.Ok(await JsonSerializer.DeserializeAsync
        <IEnumerable<GitHubBranch>>(contentStream));
```

**Review by fiyazbinhasan — COMMENTED** [review]
it looks more cleaner

File: aspnetcore/fundamentals/http-context/samples/6.x/HttpContextInBackgroundThread/BranchSyncService.cs [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+                // number of requests simultaneously.
+                using var syncTokenSource = CancellationTokenSource.CreateLinkedTokenSource(stoppingToken);
+                syncTokenSource.CancelAfter(TimeSpan.FromSeconds(30));
+                
+                var httpClient = _httpClientFactory.CreateClient("GitHub");
+                var httpResponseMessage = await httpClient.GetAsync("repos/dotnet/AspNetCore.Docs/branches", stoppingToken);
+
+                if (httpResponseMessage.IsSuccessStatusCode)
+                {
+                    using var contentStream =
```
Comment (fiyazbinhasan) [inline aspnetcore/fundamentals/http-context/samples/6.x/HttpContextInBackgroundThread/BranchSyncService.cs:36]:
```suggestion
                    await using var contentStream =
```

**Pull request merged**